### ISSUE TITLE
Error: string field contains invalid UTF-8

### ISSUE THREAD

**Issue by cmueller-tp** [Velocidex/velociraptor#2069]
Hi!
I've set up velociraptor and most of the connected clients report this error message: 
```
[DEBUG] 2022-09-10T07:31:16Z Received a post of length 1267 from <ip>:<port> (C.<id>)
[ERROR] 2022-09-10T07:31:16Z Error: string field contains invalid UTF-8

```
They're shown as connected in the UI but the hostname, OS and other information are not shown.
What other info can I provide for debugging? Can't see anything else in the logs related to the issue.

**Comment by scudette** [comment]
This looks to be like some kind of corruption of data in transit. Do you have tls stripping or a reverse proxy?

**Comment by cmueller-tp** [comment]
Wow, that was quick!
I have a reverse proxy for tls termination for the GUI, not for the client port. I know that the customer has a relatively complex network on his side, but they set an allow rule for my IP/port combination. I don't really see what else could go wrong there.

**Comment by scudette** [comment]
A good place to start is with the trouble shooting guide here https://docs.velociraptor.app/docs/deployment/troubleshooting/#debugging-client-communications

**Comment by cmueller-tp** [comment]
Thanks!
I've already looked at it. Would this error occur if the customer routed the traffic through a proxy?

**Comment by scudette** [comment]
It might but normally the connection is inside a tls connection so I don't imagine it can be corrupted. Unless they have something like a mitm proxy that decrypts the data and maybe changes it?

**Comment by cmueller-tp** [comment]
that would be really odd, but I don't really see what else it might be either...
Anyways, thanks for you help, guess I'll try tcpdump now :)

**Comment by cmueller-tp** [comment]
Oh no - we just talked about the encryption, I'm stupid...

**Comment by cmueller-tp** [comment]
Where could I add more debug info to the server? Just dumping the protobuf in that error case for example. That way I could see if it's totally messed up.

**Comment by scudette** [comment]
Did you mention the error was on the client side? 

I always spin up a VM for each engagement that I can test with. You could then make sure it works over the internet normally, and verify it's only an issue when going through VPN etc.

You could also compile from source and add various dump calls along the decryption path to see. The post is received in server/comms.go in the reader handler, so you can add a print there to see what exactly in coming in the post request

**Comment by cmueller-tp** [comment]
> I always spin up a VM for each engagement that I can test with. You could then make sure it works over the internet normally, and verify it's only an issue when going through VPN etc.

yep, that's my learning...

> You could also compile from source and add various dump calls along the decryption path to see. The post is received in server/comms.go in the reader handler, so you can add a print there to see what exactly in coming in the post request

perfect, thanks! I'll report back if I find something.

**Comment by scudette** [comment]
Here is the code in the client side that reads the data from the post

https://github.com/Velocidex/velociraptor/blob/master/http_comms/comms.go#L692

It is the first step in the decrypting process so if it's corrupted it is likely to hit an error there.

**Comment by cmueller-tp** [comment]
Ok, so I was finally able to get some more information.
I've dumped the raw data received by the server and found out that https://github.com/Velocidex/velociraptor/blob/v0.6.6/crypto/messages.go#L39 fails. Looking at the dump I can see most of the data looking completely fine but in the json there is stuff like:
```
00000360: XXXX XXXX XXXX XXXX XXXX 6334 2207 4445  xxxxxxxxxxc4".DE
00000370: 4641 554c 5428 ed05 5522 3a31 3231 2e36  FAULT(..U":121.6
00000380: 3235 2c22 5253 5322 3a35 3730 XXXX XXXX  25,"RSS":570xxxx
```
(Sorry for the mask, not sure what potential customer data might be in there)

so at 0x376 and 0x377 (values are 0xed 0x05) it's invalid - looks like the "CPU": part that comes later in other json messages as well. I guess it gets corrupted there - the DEFAULT also looks off compared to the other messages.
Is that from a proxy? The DEFAULT thing looks like it's from "withing" the encryption, otherwise there would be a full 16 byte chunk random data with a single bit flip...

**Comment by scudette** [comment]
Just to make sure, can you please indicate the exact version string on the client and server. You can get that by running `velociraptor version`

**Comment by scudette** [comment]
This is very strange - if the error occurs when parsing out the message list then it should already be verified by Velociraptor's encryption so it can not be anything done on the customer's end.

Adding a hex dump at the line you mentioned shows similar output 

```
 00000070  c7 b2 8b fa 02 1a 14 47  65 6e 65 72 69 63 2e 43  |.......Generic.C|
 00000080  6c 69 65 6e 74 2e 53 74  61 74 73 22 07 44 45 46  |lient.Stats".DEF|
 00000090  41 55 4c 54 28 03 0a a9  0c 0a 0c 46 2e 4d 6f 6e  |AULT(......F.Mon|
 000000a0  69 74 6f 72 69 6e 67 18  ba b1 a3 c1 80 83 e9 89  |itoring.........|
 000000b0  17 8a 02 8d 0c 12 09 54  69 6d 65 73 74 61 6d 70  |.......Timestamp|
 000000c0  12 03 43 50 55 12 03 52  53 53 12 0a 43 50 55 50  |..CPU..RSS..CPUP|
 000000d0  65 72 63 65 6e 74 1a d6  01 0a bd 01 53 45 4c 45  |ercent......SELE|
```

This is an encoded protobuf and DEFAULT is the log level for a message so it looks ok to me.

It is most likely to be a bug in version compatibility between the client and server versions.

**Comment by scudette** [comment]
Or maybe on second thought the part following the DEFAULT looks like a half of a JSON message - normally it should be a complete JSON object because this is the result set for the Generic.Client.Stats query

```
 000001b0  91 f4 dc 8b b4 8b fa 02  28 03 30 05 38 0e 52 b8  |........(.0.8.R.|
 000001c0  0a 7b 22 54 69 6d 65 73  74 61 6d 70 22 3a 31 36  |.{"Timestamp":16|
 000001d0  36 32 38 35 33 33 38 39  2e 35 39 36 34 38 39 37  |62853389.5964897|
 000001e0  2c 22 43 50 55 22 3a 31  2e 38 31 2c 22 52 53 53  |,"CPU":1.81,"RSS|
 000001f0  22 3a 36 33 36 32 37 32  36 34 2c 22 43 50 55 50  |":63627264,"CPUP|
 00000200  65 72 63 65 6e 74 22 3a  30 2e 30 30 30 39 39 39  |ercent":0.000999|
 00000210  39 37 32 34 38 37 32 35  32 39 34 34 36 7d 0a 7b  |9724872529446}.{|
 00000220  22 54 69 6d 65 73 74 61  6d 70 22 3a 31 36 36 32  |"Timestamp":1662|
```

So maybe it is misplaced - this is possible because the whole thing is compressed so maybe it is getting scrabled while being decompressed? A bit error early on can lead to decompression artifacts like adding parts of the message out of order etc.

**Comment by IsaAlMannaei** [comment]
I am having the same error as well. However, both the server and the client in AWS and no proxy involved. To be more specific, I only face this error when the hunts through hunt manager, but if through interactive shell it works fine. The issue is not consistent, in some cases it worked but rarely. I am still unable to fine a solution.

**Comment by scudette** [comment]
are the client and server same version?

**Comment by scudette** [comment]
Also did you build from source or are you using the distributed code? It might be some kind of issue with different versions of the protobuf compiler

**Comment by cmueller-tp** [comment]
For me it was v0.6.6 server and client (for debugging I used a self compiled version, checked out on v0.6.6).
But as I said - when connecting to the server from my laptop or a vm it works fine, just not from the customers network.

**Comment by cmueller-tp** [comment]
> This is very strange - if the error occurs when parsing out the message list then it should already be verified by Velociraptor's encryption so it can not be anything done on the customer's end.
> 
> Adding a hex dump at the line you mentioned shows similar output
> 
> ```
>  00000070  c7 b2 8b fa 02 1a 14 47  65 6e 65 72 69 63 2e 43  |.......Generic.C|
>  00000080  6c 69 65 6e 74 2e 53 74  61 74 73 22 07 44 45 46  |lient.Stats".DEF|
>  00000090  41 55 4c 54 28 03 0a a9  0c 0a 0c 46 2e 4d 6f 6e  |AULT(......F.Mon|
>  000000a0  69 74 6f 72 69 6e 67 18  ba b1 a3 c1 80 83 e9 89  |itoring.........|
>  000000b0  17 8a 02 8d 0c 12 09 54  69 6d 65 73 74 61 6d 70  |.......Timestamp|
>  000000c0  12 03 43 50 55 12 03 52  53 53 12 0a 43 50 55 50  |..CPU..RSS..CPUP|
>  000000d0  65 72 63 65 6e 74 1a d6  01 0a bd 01 53 45 4c 45  |ercent......SELE|
> ```
> 
> This is an encoded protobuf and DEFAULT is the log level for a message so it looks ok to me.
> 
> It is most likely to be a bug in version compatibility between the client and server versions.

And my messages start with json, then I have some random other message part in there and ~11-13 bytes later it seems like the original json from the first message continues.

**Comment by scudette** [comment]
Did you rebuild the protobuf files?

**Comment by cmueller-tp** [comment]
Oh, no. When do I need to rebuild them? After checking out? Or updating the server? Initially I've setup a v0.6.5-2 (I think) and then updated to v0.6.6 a day later on the server side (clients were v0.6.6 from the beginning as they were connected after the update)

**Comment by scudette** [comment]
No you do not need to rebuild them unless you are changing code - I am trying to eliminate the possibility that version of the protobuf compiler are mismatching.

**Comment by cmueller-tp** [comment]
ah, ok. The only code changes I did were adding printlns to find out what's going wrong

**Comment by scudette** [comment]
are you seeing the same issued with 0.6.5?

**Comment by cmueller-tp** [comment]
as my customer is not available during the weekend and I don't have working access to the network I can't really test atm... But maybe @IsaAlMannaei can give it a try. Otherwise I'll report back tomorrow

**Comment by scudette** [comment]
Another thing to try when you can is to turn off message compression here
https://github.com/Velocidex/velociraptor/blob/master/http_comms/comms.go#L686

Change to PackedMessageList_UNCOMPRESSED to eliminate the extra compression step - maybe we can see more clearly what is happening to the data.

**Comment by cmueller-tp** [comment]
Compression is a smart guess, that might be it. Checking it out now

**Comment by cmueller-tp** [comment]
ok so I changed https://github.com/Velocidex/velociraptor/blob/v0.6.6/http_comms/comms.go#L686 to uncompressed and also had to change https://github.com/Velocidex/velociraptor/blob/v0.6.6/utils/compression.go#L26 to `return plain_text, nil`. Now it seems to work for windows - at least on my test VM (for some reason I was able to reproduce it on one of my test vms).
And generally it looks like Windows client -> Linux server doesn't work with compressed data. Not sure why though.

**Comment by cmueller-tp** [comment]
That was a very nice hint, thanks! PR coming in a few

**Comment by cmueller-tp** [comment]
So at least on my test VM it works with this patch. I hope I can update the client automatically on the customers devices...

**Comment by cmueller-tp** [comment]
>  I hope I can update the client automatically on the customers devices...

@scudette  is that possible somehow? I think server->client works, so it would be really cool to be able to do that. They don't have an AD so it would be a bit of trouble to get that working...

**Comment by scudette** [comment]
I might have been too quick to misread the code :-( this is just the initial size of the buffer - it is allowed to grow as required. On first read I thought it was the max size of the buffer :-).

I see your patch essentially removed the ctx from the copy so maybe this is the issue - the request is already cancelled at the time we try to decompress it.  Looking at git blame this code is 3 years old :-) so it is not something that just broke.

The context comes all the way from the http handler. Maybe what is happening here is that the http request is getting cancelled prematurely. The server will then drop the connection and cancel all processing related to the request.

**Comment by scudette** [comment]
> > I hope I can update the client automatically on the customers devices...
> 
> @scudette is that possible somehow? I think server->client works, so it would be really cool to be able to do that. They don't have an AD so it would be a bit of trouble to get that working...

You dont need to update clients because messages from server to client are not compressed - only messages from client to server. It is just not work the effort compressing the relatively short VQL fragments but it is useful to compress bulk data that is send from the client to the server.

**Comment by scudette** [comment]
If we remove the ctx copy here further processing will still be cancelled just later down the track. It is worth testing if it really fixes the issue though.  Do you want to further test the patch on the server? Just sync to 0.6.6 branch so you dont need to resync it.

For testing you may try to change this code:

https://github.com/Velocidex/velociraptor/blob/master/server/comms.go#L259

to 
```go
ctx, cancel := context.WithTimeout(context.Background(), 600*time.Second)
```

This completely ignores the lifetime of the incoming request and will just take as long as needed to process.

**Comment by ekky19** [comment]
I have the same problem on windows machines. When I run an artifact doesn't work, but if through the interactive shell it works fine. Also hostname and some other client infos don't come up on th client list.

Is there any quick solution :)

**Comment by scudette** [comment]
Can you please test the server build here https://github.com/Velocidex/velociraptor/actions/runs/3041847470

It sets a hard timeout for processing so does not depend on the client maintaining the connection.

**Comment by cmueller-tp** [comment]
the server build doesn't work with v0.6.6 clients, right?
I've tried multiple different things now with the compression and the only thing that actually seems to work is turning off compression on the client side. So you were right about my "fix", it's not fixing anything.

**Comment by scudette** [comment]
Have you tried the server build I pointed to?

**Comment by cmueller-tp** [comment]
I tried it with the v0.6.6 clients and they don't work. I can also deploy test instances with the v0.6.7-dev build

**Comment by scudette** [comment]
Interesting - this should be the same as your PR as the context should not be cancelled 🤔 Are you getting it to work with your PR?

**Comment by cmueller-tp** [comment]
> Interesting - this should be the same as your PR as the context should not be cancelled thinking Are you getting it to work with your PR?

Oh no, sorry, wasn't clear on that: I was too tired yesterday, it didn't really work with that PR. It only works when disabling all compression on the client side (I think I tested the PR with that one test client that was still working because it didn't use compression, the other clients are still broken). The test build you provided seems to have changed the protobuf structure so the v0.6.6 clients are completely broken with this server version

**Comment by scudette** [comment]
Hmm that's interesting because the protobufs have not changed at all. What do you mean by completely broken with 0.6.6 clients - what is the error?

I mean we can turn off compression as an option but it is very strange why that would be the issue.

**Comment by scudette** [comment]
Can you please test with 0.6.5? I'm concerned it could be a regression. 

Thanks again for working through this issue...

**Comment by cmueller-tp** [comment]
Ok, thats ultra strange. Yesterday I got some message that the format of the buffer is not valid, then the frontend didn't load and I got multiple tls error logs the velociraptor server log. That's why I assumed the protobuf version didn't match.
Now when trying to reproduce it to give you the exact error message the server just dies randomly (3x within like 5 minutes now), I still get the `Error: string field contains invalid UTF-8` message here and there and the hunt I started still doesn't give me hostname and the clients are running for ever.
Ah, and now I found the message I was looking for (`Error: proto: cannot parse invalid wire-format data
`). A bit more output:
```
[DEBUG] 2022-09-13T09:30:50Z Received a post of length 835 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:51Z Received a post of length 1187 from <ip>:<port> (<clientid>)
[ERROR] 2022-09-13T09:30:51Z Error: string field contains invalid UTF-8
[DEBUG] 2022-09-13T09:30:51Z Received a post of length 690 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:52Z Received a post of length 1027 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:52Z Received a post of length 690 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:52Z Received a post of length 803 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:53Z Received a post of length 819 from <ip>:<port> (<clientid>)
[ERROR] 2022-09-13T09:30:53Z Error: proto: cannot parse invalid wire-format data
[DEBUG] 2022-09-13T09:30:53Z Received a post of length 690 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:54Z Received a post of length 690 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:54Z Received a post of length 1283 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:54Z Received a post of length 1091 from <ip>:<port> (<clientid>)
[ERROR] 2022-09-13T09:30:54Z Error: string field contains invalid UTF-8
[DEBUG] 2022-09-13T09:30:54Z Received a post of length 690 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:54Z Received a post of length 851 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:54Z Received a post of length 851 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:54Z Received a post of length 1011 from <ip>:<port> (<clientid>)
[DEBUG] 2022-09-13T09:30:55Z Received a post of length 1011 from <ip>:<port> (<clientid>)
```

**Comment by scudette** [comment]
The invalid wire format data is same as the utf8 error. It is because the data seems to be corrupted. 

When you say it crashed do you have a stack trace? The server has a panic guard normally that will write crash logs too the logs directory is it crashes

**Comment by cmueller-tp** [comment]
Assuming you mean `/opt/velociraptor/logs/` no, it's empty. Not log in the server output either, just dies silently.

**Comment by scudette** [comment]
Can you try a 0.6.5 server please?

**Comment by cmueller-tp** [comment]
> Can you try a 0.6.5 server please?

yep, still the same problem, doesn't seem to be a regression...
Sorry, didn't see your request earlier. And no problem, I really wanna try out velociraptor. Unlucky I get these issues with the first try :D

**Comment by scudette** [comment]
Ok I think I tried another thing - My theory was that maybe the proxy limits the size of POST requests so I tried to simulate truncation of the message and it shows similar problems with decoding protobufs. Is it possible that the POST size is limited by some network proxy? We normally try to limit the size of POST to about 5-10mb. 

This kind of issue manifests in very small packets getting through but larger packets are corrupted. But maybe this is not the same issue because the corruption happens in the decryption code.

The error looks like this when the post is truncated (i.e. it rejects it in the decryption code.
```
[DEBUG] 2022-09-13T03:43:10-07:00 Unable to decrypt body from 127.0.0.1:61721: proto: cannot parse invalid wire-format data
www.velocidex.com/golang/velociraptor/crypto/client.(*CryptoManager).Decrypt
        /home/mic/projects/velociraptor/crypto/client/manager.go:212
www.velocidex.com/golang/velociraptor/server.(*Server).Decrypt
        /home/mic/projects/velociraptor/server/server.go:254
www.velocidex.com/golang/velociraptor/server.readWithLimits
```

I can't figure out how a packet can be correctly decrypted but still be corrupt.

**Comment by scudette** [comment]
Maybe this is not the issue - looking at your log lines about the POST sizes are really small .

**Comment by cmueller-tp** [comment]
> I can't figure out how a packet can be correctly decrypted but still be corrupt.

As long as it's 16 byte aligned decryption should work without problems (I think I saw aes-128-cbc is used for encryption?). I'd recommend to use `aes-256-gcm` anyways. That way you have authentication in the crypto layer. CBC is not commonly used anymore as it has many problems like missing authentication, you can bitflip specific bits in known plaintext attacks without compromising the entire message (only one block) and so on.

But anyways, I think POST size should not be the problem. Oh, you just mentioned it... 

Taking a step back:
I was able to reproduce the issue on my local test VM. I know there is no proxy in the way. I can dump the compressed message and even decode it with openssl zlib and I still get corrupted messages. With Linux clients there is no problem. 
 My guess therefor is that the Windows client does something different to the Linux client.
What I really don't understand is why. the zlib implementation is pure go code without external dependencies, why would it work differently on Windows?

**Comment by scudette** [comment]
Re encryption we use hmac to ensure message integrity so it should not be possible to mutate the message in transit.

I think everything is pointing to some incompatibility as you say between windows and linux compression? as unlikely as it sounds.  Does the issue go away 100% when you turn compression off?

**Comment by scudette** [comment]
Also FWIW zip has checksums so it should usually detect corruption in the decompression step.

**Comment by cmueller-tp** [comment]
Mhh, I'm out of ideas tbh. The Windows->Linux thing sounds soooo unlikely... I'll test again with compression off. Btw, there are multiple calls to `utils.Compress` that are not guarded with 
```
if compression == crypto_proto.PackedMessageList_ZCOMPRESSION {}
```
I think that's why I had to disable the compression in 2 places (for example `http_comms/comms.go:763`)

**Comment by scudette** [comment]
Ok if it turns out that compression is somehow the issue here we can make it optional.

**Comment by cmueller-tp** [comment]
This issue seems to be resolved without compression. Not sure if #2079 may be connected to my changes, but I don't think so.

**Comment by scudette** [comment]
Ok this is a very curious bug for sure! Just adding more information here....

During work on #2081 I noticed that in this code

https://github.com/Velocidex/velociraptor/blob/899fc6eef65ba8c030d2db757314d9554f365138/crypto/messages.go#L26-L43

I can see the exact same error you get when the message is really compressed, but the self.Compression indicates it is not... so we then attempt to parse the protobuf from a compressed blob which fails. The data looks very similar to what you had.

I just can not figure out why the compression here is marked off by the client but it is really compressed. Can you please add the following code so we can see what is happening:

```go
                message_list := &crypto_proto.MessageList{}
		err := proto.Unmarshal(raw, message_list)
		if err != nil {
                        utils.Debug(self)
			return errors.WithStack(err)
		}
```

**Comment by scudette** [comment]
The latest PR should allow for compression to be disabled on the client - this will require a client config update though 

Can you please test these CI builds to see if the issue is resolved?
https://github.com/Velocidex/velociraptor/actions/runs/3045396057

**Comment by scudette** [comment]
Looking more closely at this I now believe this is actually a manifestation of #2064 

The client has a ring buffer file which queues up messages going out to the server. Normally the file is controlled via an internal FileRingBuffer object which serializes access to it by use of a mutex. The assumption is that there is only one manager of the ring buffer file and so there is no need to rely on filesystem locking.

However due to #2064 by accident two communicators were started and two instances of the file managers were created. Each one is supposed to have exclusive ownership of the file but when two managers are created they do not lock the file between them.

This leads to file corruption as both managers are trying to queue messages on the file. This is why we see some messages overwriting the middle of other messages.

Release 0.6.6-2 fixes the issue but this requires a client upgrade.

It should be possible to do a remote update via Admin.Client.Upgrade as the ring buffer only manages messages from the client to the server.

**Comment by scudette** [comment]
![image](https://user-images.githubusercontent.com/3856546/190154801-2386d710-936e-4b99-ac22-59c626ca2553.png)

This is exactly what we are seeing - the first part is the stats JSON object which has rows every 10 seonds. Then right in the middle we get the stats log message ( Sending response part 1028 1.1 kB), then it goes on with the json object as if nothing happened. Clearly the messages are getting trashed over each other.

**Comment by cmueller-tp** [comment]
Yep, that does the trick. Thanks a lot!

**Comment by scudette** [comment]
This bug is fixed in 0.6.6-2 i think right?

Please reopen if you are still experiencing issues
